Debian: Fix starting/stopping of units on package (un)install

martinpitt opened the pull request:
This syncs relevant changes from https://anonscm.debian.org/cgit/collab-maint/cockpit.git to our Debian packaging in the upstream tree.

This adds a new dh-systemd build dependency, which needs to be echoed three times in `test/images/scripts/*` (this is issue #3298) and thus requires building new test base images. Thus the initial auto-triggered integration tests for debian and ubuntu will fail, and we'll need to tell the bot to refresh them first.

martinpitt commented:
bot: Image refresh for debian-unstable

martinpitt commented:
bot: Image refresh for ubuntu-1604

martinpitt commented:
bot: Image refresh for debian-8

petervo commented on tools/debian/control:
```diff
@@ -77,7 +77,6 @@ Package: cockpit-docker
 Architecture: all
 Depends: ${misc:Depends},
          docker.io (>= 1.3.0) | docker-engine (>= 1.3.0),
-         python,
```
I think the docker package does use python in cockpit.spawn commands.

martinpitt replied:
Nothing in the shipped package seems to:
```sh
$ dpkg-deb -x cockpit-docker_0-1_all.deb c
$ grep -r python c
# (no results)
```
Although there clearly are hits in `git grep python {dist,pkg}/docker`, but neither `cockpit-atomic-storage` nor `storage.jsx` is shipped in the binary docker package. Should they be?

petervo replied:
Storage.jsx is built into docker.js by webpack. and I believe that is distributed on debian. However, it is true, there is no such thing as debian atomic and it only used for dealing with storage on atomic. So it's probably ok to leave out. My only concern would be that future devs will see use python in some other way in the docker package and forget to update the requirements because technically it was already in use in the docker code. But i'll leave it up to you.

martinpitt replied:
TBH I'd rather rely on the integration tests to spot this (and then move to Python 3) No other cockpit binary depends on python 2, and I'd rather not introduce a new Debian package that uses Python 2. Debian and Ubuntu has worked on getting rid of py 2 for many years.

mvollmer replied:
> However, it is true, there is no such thing as debian atomic and python is only used for dealing with storage on atomic.

The cockpit-atomic-storage script is used everywhere.  Depending on what is installed, it will let you manage the storage or not.  In all cases it is used to monitor the storage usage of images.

Thus, it always needs to run, even on Debian and Ubuntu.

Let's make sure it works with Python 3.

Does it need to work with both Python 2 and 3?  I am not sure.  The "atomic" command itself supports that for some reason.

mvollmer replied:
> Nothing in the shipped package seems to [refer to python]:

It's probably in a gzipped file.

stefwalter replied:
@martinpitt see ```pkg/docker/storage.jsx``` and ```pkg/docker/cockpit-atomic-storage``` if you want to make this work well with python 3.

martinpitt commented:
Semaphore failed on three pyflakes issues, which are fixed in PR #5701. But let's wait for the other tests first, I'll need to force-push anyway to update the image symlink UUIDs once the bot generated them.

cockpituous commented:
Image creation for ubuntu-1604 in process on cockpit-11.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-ubuntu-1604-267e615827057276067a4e1b50d5cd7147317fec/

cockpituous commented:
Image creation for debian-unstable in process on cockpit-9.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-debian-unstable-267e615827057276067a4e1b50d5cd7147317fec/

cockpituous commented:
Image creation failed.

cockpituous commented:
Image creation done: https://github.com/cockpituous/cockpit/commits/refresh-ubuntu-1604-267e615827057276067a4e1b50d5cd7147317fec

martinpitt commented:
Pushed the ubuntu-1604 image refresh commit. debian-unstable apparently isn't buildable right now due to some library transition, and debian-8 first needs PR #5716 to land until the bot can rebuild them.

dperpeet commented:
Needs rebase because of the [new ubuntu image](https://github.com/cockpit-project/cockpit/commit/d9ffac3ca480e27f652deab5e9bf0a81e7883fc0) that was merged. It was only a refresh, nothing changed, so I don't think this one needs to be recreated.

martinpitt commented:
@dperpeet: Rebased and force-pushed.

cockpituous commented:
Image creation for debian-8 in process on cockpit-7.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-debian-8-8f2e58bfec92544806070f49bd2ab12588cf69a5/

stefwalter commented on tools/debian/control:
```diff
@@ -70,26 +74,26 @@ Description: Cockpit deployment and developer guide
+Architecture: all
 Depends: ${misc:Depends},
          docker.io (>= 1.3.0) | docker-engine (>= 1.3.0),
-         python
```
We need python for the pkg/docker/storage.jsx stuff.

martinpitt replied:
See discussion with Peter above. The Debian package does not ship the bits that call Python (allegedly it's only for interacting with Atomic), and I really want to avoid introducing a new Python 2 dependency into Debian/Ubuntu particularly if it's not actually required.

stefwalter replied:
Not sure which Debian package you're referring to, but I definitely see this installed when I use our http://cockpit-project.org/running.html packages on Debian:

```
root@unassigned-hostname:~# grep VERSION /etc/os-release 
VERSION_ID="8"
VERSION="8 (jessie)"
root@unassigned-hostname:~# zgrep python /usr/share/cockpit/docker/* 
/usr/share/cockpit/docker/docker.min.js.gz:                l.hidden || i ? l.hidden && i && (i.close("closed"), i = null) : i = l.spawn([ "python", "-c", u, "monitor" ], {
/usr/share/cockpit/docker/docker.min.js.gz:                var r = l.spawn([ "python", "-c", u, s ].concat(i), {
/usr/share/cockpit/docker/docker.min.js.gz:                var n = l.spawn([ "python", "-c", u, "reset-and-reduce" ], {
```

stefwalter replied:
And when running, I definitely see it in use:

```
root@unassigned-hostname:~# ps -xa | grep python
23363 ?        S      0:00 python -c #! /usr/bin/python  # This utility fills in the gaps between the 
```

martinpitt replied:
Right, sorry about that. Just force-pushed, that commit is gone now.

cockpituous commented:
Image creation done: https://github.com/cockpituous/cockpit/commits/refresh-debian-8-8f2e58bfec92544806070f49bd2ab12588cf69a5

stefwalter commented on tools/debian/control:
```diff
@@ -28,24 +30,26 @@ Build-Depends: debhelper (>= 9),
+         cockpit-bridge (>= ${source:Version}),
+         cockpit-bridge (<< ${source:Version}.1~),
+         cockpit-ws (>= ${source:Version}),
+         cockpit-ws (<< ${source:Version}.1~),
```
I believe the cockpit meta package should enforce that identical versions of the subpackages are installed. If not, then why not? If someone wants to mix and match parts of cockpit, then the cockpit meta package is something that should be removed.

martinpitt replied:
That's precisely what this does, what do you mean? This is the usual way of specifying dependencies from an `Architecture: all` to an `Architecture: any` package, as the latter can get "binary NMUs", i. e. get rebuilt with a `b1` suffix on library transitions; but that  doesn't rebuild the `Arch: all` packages, so `${binary:Depends}` doesn't work in this case.

stefwalter replied:
Aha ok. Misunderstood.

stefwalter commented on tools/debian/control:
```diff
@@ -70,26 +74,26 @@ Description: Cockpit deployment and developer guide
+Architecture: all
+Depends: ${misc:Depends},
+         cockpit-bridge (>= ${source:Version}),
+         cockpit-bridge (<< ${source:Version}.1~),
```
There's a certain minimum cockpit-bridge that we need to require here. We just haven't gotten around to it. See https://github.com/cockpit-project/cockpit/pull/5277 Perhaps that minimum version requirement can be handled in another pull request.

martinpitt replied:
What do you mean? The inter-package dependencies are set up to enforce that all cockpit* packages have the same version, except for the slight binNMU version suffixes (see above).

stefwalter replied:
It's fine for Debian/Ubuntu packages to keep this simple for now. 

When the cockpit meta package is installed it enforces that they all remain in sync. But going forward there's definitely the ability for various packages like cockpit-networkmanager or cockpit-storaged to be upgraded separately and have their version numbers out of sync when installed individually. There's a certain base minimum version number of the bridge that's required. You can see this in the RPM spec file.

martinpitt replied:
Ah, I see what you mean, I thought you were saying that the depends isn't strict enough. Maybe we can take `required_base` out of `cockpit.spec` in the future and into something which both the spec file and debian/rules read?

stefwalter replied:
Yes lets do something like that in the future. Bonus points for using the version number in manifest.json in the package itself. That way we don't have to duplicate the information.

martinpitt replied:
I did that now in PR #5845 .

stefwalter commented on tools/debian/control:
```diff
@@ -103,11 +107,20 @@ Depends: ${misc:Depends},
-         ${shlibs:Depends},
-         cockpit-bridge (= ${binary:Version}),
+         cockpit-bridge (>= ${source:Version}),
+         cockpit-bridge (<< ${source:Version}.1~),
```
Ditto for above mimimum cockpit-bridge requirement.

martinpitt commented:
Force-pushed with updated debian-8 image. debian-unstable is still unbuildable (just checked in a chroot that `pcp` is uninstallable -- maybe we should move to Debian testing?)

martinpitt commented:
Confusingly @mvollmer's review doesn't appear at all here any more, so quoting from email:

> The cockpit-atomic-storage script is used everywhere.

This isn't being shipped in `cockpit-docker`  nor in any other binary (Atomic is not a Debian thing).

> It's probably in a gzipped file.

Ah, ok, I see! I'll drop that commit (which removes the python dep) for now. I filed issue #5719 as a reminder.

mvollmer commented:
> Confusingly @mvollmer's review doesn't appear at all here any more

You can dig out via one of the "Show outdated" buttons.

mvollmer commented:
> This isn't being shipped in cockpit-docker nor in any other binary (Atomic is not a Debian thing).

There is no file named `cockpit-atomic-storage` in the binary packages, but its content is shipped as part of docker.js.gz.

We use the JavaScript module loader to also load raw files, such as some short bash snippets and the pretty large cockpit-atomic-storage program.  For distribution, webpack puts them into a single file together with all other JavaScript modules.

I think we do this mainly so that the static assets of a component (such as cockpit-docker) can be relocated easily, for example into a container.

stefwalter commented:
This test failure looks new:

```
# testCommandline (check_connection.TestConnection)
#
Wrote TestConnection-testCommandline-FAIL.png
Journal extracted to TestConnection-testCommandline-10.111.117.93-FAIL.log
not ok 48 testCommandline (check_connection.TestConnection) duration: 15s
Traceback (most recent call last):
  File "./verify/check-connection", line 214, in testCommandline
    self.assertFalse(m.execute("ls /etc/cockpit/ws-certs.d/* || true"))
AssertionError: '/etc/cockpit/ws-certs.d/0-self-signed.cert\n' is not false
```

martinpitt commented:
FTR, I locally fixed the conflicts with the recently landed -dashboard splitout, but I can reproduce the `testCommandline` failure locally, so I won't push yet.

martinpitt commented:
I know why that test case fails: I added `dh_systemd_start -p cockpit-ws --no-start --restart-after-upgrade cockpit.service` with the intention to restart cockpit.service on package upgrade if it was running, but not on initial installation (only `cockpit.socket` should be started then). This is a bug in `dh_systemd_start`, it generates code which starts cockpit.service on initial installation, and thus `/etc/cockpit/ws-certs.d/0-self-signed.cert` gets generated in the test VM already. I'll replace this with manual and correct postinst code.

martinpitt commented:
Re-pushed with:
 - rebasing to master (conflicts due to the -dashboard splitout)
 - a workaround for that dh_systemd bug (https://bugs.debian.org/805878 -- it got fixed recently, but we cannot rely on that for stable releases yet)
 - adding missing `Replaces:` to the new -dashboard package, to avoid upgrade failures on file conflicts

stefwalter commented:
We're rolling another release due to the Ubuntu/Debian cockpit-shell -> cockpit-system problems. I'm merging everything from here except bc4786828f44b65e40f0a69999e6584ee276a3e1 and the image rebuilds. Those need another pull request due to conflicts and the issues with debian-testing vs. debian-unstable.

martinpitt commented:
WIPing this for now, as it needs to (or at least should) wait on PR #5726 . I rebased to master and added a bunch more fixes, plus calling lintian to ensure that these kinds of errors don't come back.

martinpitt commented:
Force-pushed. I rebased to current master,  and are currently running an image build and tests against debian-testing. I'll ask the bot to do the image updates for ubuntu-1604 and debian-8 again, to parallelize this.

martinpitt commented:
bot: Image refresh for ubuntu-1604

martinpitt commented:
bot: Image refresh for debian-8

cockpituous commented:
Image creation for ubuntu-1604 in process on cockpit-10.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-ubuntu-1604-63dfd17a296478d7be980cd86f3ee384a48ef713/

cockpituous commented:
Image creation for debian-8 in process on cockpit-8.
Log: http://fedorapeople.org/groups/cockpit/logs/refresh-debian-8-63dfd17a296478d7be980cd86f3ee384a48ef713/

cockpituous commented:
Image creation done: https://github.com/cockpituous/cockpit/commits/refresh-ubuntu-1604-63dfd17a296478d7be980cd86f3ee384a48ef713

cockpituous commented:
Image creation done: https://github.com/cockpituous/cockpit/commits/refresh-debian-8-63dfd17a296478d7be980cd86f3ee384a48ef713

martinpitt commented:
Note: I didn't upload my updated debian-testing image yet, so this will fail. This is still WIP.

martinpitt commented:
bot: Image refresh for debian-testing

martinpitt commented:
bot: Image refresh for debian-testing

martinpitt commented:
I narrowed the scope of this to just the dh-systemd bits, so that we can land this quickly and avoid eternally-conflicting image rebuilds. I'll split out the lintian/licensing fixes into a separate PR - that won't need new images, and needs some more work anyway.

Keeping this WIP until the bot rebuilds the debian-testing image.

martinpitt commented:
The bot didn't like me wrt. rebuilding debian-testing, so I built/uploaded my own again.

martinpitt commented:
The `check-basic` hang is real, investigating.

martinpitt commented:
Fix for hanging `check-basic` pushed.

martinpitt commented:
`check-connection TestConnection.testSocketPort` had the same problem (assuming that cockpit.socket does not start at boot, which is a Fedora specific systemd preset assumption), fixed.

martinpitt commented:
The check-networking failure in fedora-25 is quite obviously unrelated (this PR only changes `tools/debian/` and `check-{loopback,connection}`), so I think we can ignore this and not put additional load on the busy test machines.

martinpitt commented:
The `test/verify/check-connection TestConnection.testSocketPort` failure on ubuntu-1604 is pretty sticky. Instead of "Connection refused" curl fails with "gnutls_handshake() failed: Error in the pull function.". If I add a `time.sleep(3)` before the curl then it's failing in the expected way. This is apparently some race condition with stopping the socket and `systemctl daemon-reload`.

This could be robustified by a retry loop, or figuring out what exactly to wait for, or simply by accepting a different error message. I chose to do the latter. This will reset all the test results again, but I already retried the `ubuntu-1604` one three times and this is getting annoying both for this PR and the future as well.

martinpitt commented:
I re-ran the entire test suite locally again, and found that `test/verify/check-loopback -stv TestLoopback.testBasic` would fail like this:
```
Traceback (most recent call last):
  File "test/verify/check-loopback", line 38, in testBasic
    m.wait_for_cockpit_running()
  File "/home/martin/upstream/cockpit/test/common/testvm.py", line 1387, in wait_for_cockpit_running
    self.execute(script=WAIT_COCKPIT_RUNNING)
  File "/home/martin/upstream/cockpit/test/common/testvm.py", line 382, in execute
    ret = select.select(rset, wset, [], 10)
  File "/home/martin/upstream/cockpit/test/common/testvm.py", line 83, in handle_timeout
    raise Exception(self.error_message)
Exception: Timeout while waiting for cockpit to start

ADDRESS: 10.111.115.77
Press RET to continue... 
```

which on itself is a race condition to fix. However, after pressing Enter, I get an eternal hang, apparently when the test is trying to run cleanups (but no helpful output, just the repetition of the traceback). This might explain the hanging tests.

martinpitt commented:
When it hangs on cleanup, I see these two processes lingering (aside from qemu):
```
martin   16235  0.0  0.0  72864  6208 pts/0    S+   22:14   0:00 ssh -p 22 -i /home/martin/upstream/cockpit/test/common/identity -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o BatchMode=yes -M -o ControlPath=/home/martin/upstream/cockpit/test/tmp/ssh-%h-%p-%r-16191 -o LogLevel=ERROR -l root 10.111.115.77 /bin/bash -c 'echo READY; read a'
martin   16244  0.0  0.0  72864  6352 pts/0    S+   22:14   0:00 ssh -p 22 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -o BatchMode=yes -o ControlPath=/home/martin/upstream/cockpit/test/tmp/ssh-%h-%p-%r-16191 -l root 10.111.115.77 sh -s -x
```
When ^Cing the test process, the interesting part of the backtrace...
```
  File "/usr/lib64/python2.7/unittest/case.py", line 422, in doCleanups
    function(*args, **kwargs)
  File "/home/martin/upstream/cockpit/test/common/testlib.py", line 464, in <lambda>
    self.addCleanup(lambda: machine.kill())
  File "/home/martin/upstream/cockpit/test/common/testvm.py", line 1227, in kill
    self.disconnect()
  File "/home/martin/upstream/cockpit/test/common/testvm.py", line 118, in disconnect
    self._kill_ssh_master()
  File "/home/martin/upstream/cockpit/test/common/testvm.py", line 270, in _kill_ssh_master
    self.ssh_process.wait()
  File "/usr/lib64/python2.7/subprocess.py", line 1392, in wait
    pid, sts = _eintr_retry_call(os.waitpid, self.pid, 0)
  File "/usr/lib64/python2.7/subprocess.py", line 476, in _eintr_retry_call
    return func(*args)
```
... reveals that testvm.py's `_kill_ssh_master()` is apparently hanging forever, as the `wait()` call has no timeout. Unlike Python 3's version, Python 2's `wait()` does not have a timeout, so this needs to be done manually. I'll add a timeout/kill after timeout to this to avoid the eternal hang, and look what's causing the original test failure.

martinpitt commented:
Curiously, adding a `time.sleep(3)` after `systemctl stop cockpit.socket cockpit.service` makes the test pass reliably. So this sounds similar to the above weird curl error message ("gnutls_handshake() failed: Error in the pull function." instead of "connection refused") -- it seems that in systemd 229 there might be some race condition with stopping sockets?

Indeed when I change the line to
```
        m.execute("systemctl stop cockpit.socket cockpit.service; netstat -avptn; systemctl status cockpit.socket || true")
```
I see
```
tcp6       0      0 :::9090                 :::*                    LISTEN      1332/(agetty)   
```
thus port 9090 is not really closed after stopping the unit. I added a loop that waits for the socket to really go down. This is also a better/more robust fix than the previous curl error message workaround, so I reverted that as well and replaced it with the same netcat wait loop. This makes the commits in this PR simpler and more consistent.

martinpitt commented:
FTR: debian-testing seems to hang for a long time, but I cannot reproduce this locally at all:
```
ok 128 testBasic (check_loopback.TestLoopback) duration: 25s
# TESTS PASSED [3464s on donald]
```

martinpitt commented:
I've been trying to reproduce that hang on the production infra and locally, both on the host and in the cockpit-verify docker container, to no avail. So in order to not stall this image refresh PR eternally, I now reverted the previous hack in `check-connection` and `check-loopback` and just disable `cockpit.socket` after installing the debs. Right now the test suite expects to be able to change the configuration of the unit files and cockpit to not run until it calls `start_cockpit()`, so it's easier and more robust all around to keep that behaviour. This now seems to have addressed the eternal hangs.

I keep the previous patch at https://gist.github.com/martinpitt/d3889e1bba34c80e5681505b4af01cde in case we want to revisit that in the future.

martinpitt commented:
I believe this is ready for merge now. No hangs, and the remaining failures are just the usual networking/realm suspects (and this doesn't change anything on rhel/fedora anyway).
